File: solutions/go/luhn/1/luhn.go
package luhn

import "unicode"

func Valid(id string) bool {
	var digits []int

	// Step 1: Clean input and validate characters
	for _, r := range id {
		if r == ' ' {
			continue
		}
		if !unicode.IsDigit(r) {
			return false
		}
		digits = append(digits, int(r-'0'))
	}

	// Must contain at least two digits
	if len(digits) <= 1 {
		return false
	}

	sum := 0
	double := false

	// Step 2: Traverse from right to left
	for i := len(digits) - 1; i >= 0; i-- {
		d := digits[i]

		if double {
			d *= 2
			if d > 9 {
				d -= 9
			}
		}

		sum += d
		double = !double
	}

	// Step 3: Check divisibility by 10
	return sum%10 == 0
}

File: solutions/python/luhn/1/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num

    def valid(self):
        # Remove spaces
        num = self.card_num.replace(" ", "")
        
        # Must be more than 1 digit and contain only digits
        if len(num) <= 1 or not num.isdigit():
            return False
        
        total = 0
        reverse_digits = num[::-1]
        
        for i, digit in enumerate(reverse_digits):
            n = int(digit)
            
            # Double every second digit (1-based index from right → i % 2 == 1)
            if i % 2 == 1:
                n *= 2
                if n > 9:
                    n -= 9
            
            total += n
        
        return total % 10 == 0